### ISSUE TITLE
Stanek : charge function bug ,and general inconsistency.

### ISSUE THREAD

**Issue by borisflagell** [danielyxie/bitburner#3650]
NetscriptStanek.chargeFragment (x,y) got an "undefined variable" issue.
![image](https://user-images.githubusercontent.com/104104269/168390745-138b1cde-eafa-4167-bf66-fb94b7f266ca.png)

It can theorically be easily fixed by either : 
- using workerScript.scriptRef.threads as the variable displayed by the message.
- adding a return value to IStaneksGift.charge().

But there's an inconsistency in Stanek tutorial modal that irks me. It states that : 

> When the  function finishes executing, the fragment's charge levels will be raised by an amount corresponding to the number of threads that were used. Note that it is no more effective to charge a fragment many  times with few threads than to charge few times with many threads, so there is no need to distribute  charging jobs across multiple scripts. 

But it's not actually true, as : 
- charging a fragment at least once with a higher thread count will increases noticibly its effect. 
- there's actually a soft cap passed which increasing thread count will decrease stanek efficiency (up to -6% of its optimal value)

So the best strategy is to actually run a single script at max available thread; up to about 2 millions thread. Then, it become more efficient to run several script in parallel, using at most 2 millions thread.
And there's little way for the player to legitimately understand that. Because, 
A/ the tutorial modal is misleading. 
B/ the highestCharge and numCharge variables that compound the total fragment's charge count are always obfuscated behind it.
C/ the charge and effect formula are never disclosed. (probably desirable though)

It feels to me either the charge and calculateEffect formula are needlessly complicated, either the player needs to be given more information.

I'm betting on the first, as the CalculateEffect() is both outdated (use old variable name, not consistent with 1.6.3 change) and clunky (use a needlessly complicated ` * (Math.log(1.8) * 100))` instead of just `25.5` as a vestige of the heat system)

**Comment by hydroflame** [comment]
```ts

(function () {
  const simulate = (charges: number[]): number => {
    const fragment = {
      numCharge: 0,
      highestCharge: 0,
    };
    const effect = (h: number, n: number): number => (Math.log(h + 1) / 60) * Math.pow((n + 1) / 5, 0.07);
    const charge = (t: number): void => {
      if (t > fragment.highestCharge) {
        fragment.numCharge = (fragment.highestCharge * fragment.numCharge) / t + 1;
        fragment.highestCharge = t;
      } else {
        fragment.numCharge += t / fragment.highestCharge;
      }
    };
    for (const c of charges) {
      charge(c);
    }
    return effect(fragment.highestCharge, fragment.numCharge);
  };
  const defaultCharges = [5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5, 5e5];
  const def = simulate(defaultCharges);
  const run = (charges: number[]): void => {
    for (let i = 0; i < 1000; i++) {
      const n = simulate(shuffle(charges));
      if (n !== def) {
        console.log(`${n - def}`);
      }
    }
  };

  run(defaultCharges);
  run([2e6, 2e6, 1e6, 1e6]);
  run([4e6, 2e6]);
  run([3e6, 3e6]);
  run([6e6]);
})();``` fair enough

**Comment by borisflagell** [comment]
![image](https://user-images.githubusercontent.com/104104269/169673419-a39ce2a9-3cff-4040-87a3-bca697b4b53b.png)
 Here's an illustration : 
x-axis is max charge variation.
y-axis is the resulting base effect's multiplier.

each curve correspond to a total charge value.

TL;DR : 
Total_charge is a no-brainer. The bigger the better. And it won't change the shape of the curve.
Max_charge is best when kept at around 2M; Its most efficient value.

**Comment by borisflagell** [comment]
Fixed by hydro some month ago.
